### PR TITLE
Add support for AC budget API

### PR DESCRIPTION
**Pull request by tohskai**
Inspired by the blogpost:
https://pytorch.org/blog/activation-checkpointing-techniques/

### CODE REVIEW

File: torchtitan/config/job_config.py [hunk 1]
```diff
@@ -527,7 +527,7 @@ class Checkpoint:
 
 @dataclass
 class ActivationCheckpoint:
-    mode: Literal["selective", "full", "none"] = "selective"
+    mode: Literal["selective", "full", "budget", "none"] = "selective"
```
Comment (wwwjn) [inline torchtitan/config/job_config.py]: I would suggest use name "memory_buget" because it's setting budget for memory usage

**Review by wwwjn — COMMENTED** [review]
I read the blog and find the memory budget idea is cool. Have you try out the implementation on some model (eg, llama3) with `torch.compile`?  I'm curious does it works end to end and if the performance are better

**Comment by tohskai** [comment]
> I read the blog and find the memory budget idea is cool. Have you try out the implementation on some model (eg, llama3) with `torch.compile`? I'm curious does it works end to end and if the performance are better

I haven't done runs on llama3, but on our benchmarks on it showed significant improvements over regular SAC.  This is why I wanted to upstream this :)
image
image

But our model is quite different, so it's totally reasonable to see less gains.

File: torchtitan/config/job_config.py (hunk 1 above)
Comment (tohskai) [reply torchtitan/config/job_config.py]: i like it being one word as other modes, maybe "membudget"?

**Comment by wwwjn** [comment]
Thanks for sharing! We would love to see more verifications - eg, correctness and loss curves , and performance analysis on titan supported models (llama3, etc)

cc @soulitzer  for reviewing

File: torchtitan/distributed/activation_checkpoint.py [hunk 2]
```diff
@@ -169,13 +169,17 @@ def apply_ac(
...
         )
-    for layer_id, transformer_block in model.layers.named_children():
-        transformer_block = _apply_ac_to_transformer_block(
-            transformer_block,
-            ac_config,
-            base_fqn=f"layers.{layer_id}",
-            save_list=save_list,
-        )
-        model.layers.register_module(layer_id, transformer_block)
+    if ac_config.mode == "budget":
```
Comment (tianyu-l) [inline torchtitan/distributed/activation_checkpoint.py]: Under this option, we need to assert `model_compile_enabled` is True.

File: torchtitan/config/job_config.py (hunk 1 above)
Comment (soulitzer) [reply torchtitan/config/job_config.py]: I prefer `memory_budget` as well. Although it doesn't fit the pattern of "one word", I think we should separate the two words and avoid abbreviation for readability.

File: torchtitan/config/job_config.py [hunk 3]
```diff
@@ -555,6 +555,10 @@ class ActivationCheckpoint:
     Whether to stop recomputing early when all activations have already been
     rematerialized.
     """
+    budget_ac_option: float = 1.0
```
Comment (soulitzer) [inline torchtitan/config/job_config.py]:
```suggestion
    activation_memory_budget: float = 1.0
```

File: torchtitan/config/job_config.py [hunk 4]
```diff
@@ -555,6 +555,10 @@ class ActivationCheckpoint:
...
+    """
+    Limit memory budget for AC from 0-100%. Works only with compile enabled.
```
Comment (soulitzer) [inline torchtitan/config/job_config.py]: For the doc here, we could do something like:
```
When mode is set to "memory_budget", this value determines how much
partitioner in the compiler should trade off compute for memory.
0.0 corresponds to the activation memory from applying
activation checkpointing to the full compiled region, and 1.0 corresponds to
the activation memory from the default runtime-optimized strategy. 
```
(taken from `torch/_functorch/config.py`)

**Review by tianyu-l — CHANGES_REQUESTED** [review]
> I haven't done runs on llama3, but on our benchmarks on it showed significant improvements over regular SAC. This is why I wanted to upstream this :)

I agree with @wwwjn that
> We would love to see more verifications - eg, correctness and loss curves , and performance analysis on titan supported models (llama3, etc)

Please refer to https://github.com/pytorch/torchtitan/blob/main/CONTRIBUTING.md#proof-of-value

**Comment by tohskai** [comment]
@wwwjn @soulitzer @tianyu-l 

Should this support selection of `activation_memory_budget_solver` and `activation_memory_budget_runtime_estimator`? What about `visualize_memory_budget_pareto`? I found them to be useful, and the discoverability is low, but given that this is unstable api and potentially feature overload, I would prefer to hear your opinion.  

https://github.com/pytorch/pytorch/blob/main/torch/_functorch/config.py#L147-L169

**Comment by tohskai** [comment]
I ran `models/llama3/train_configs/llama3_8b.toml` on 8xH100:

image

image

image

image

Would that suffice for Proof of Value?

**Comment by tohskai** [comment]
I addressed your comments, rebased to avoid conflicts and added the other parts of the api, but I am totally okay with reverting it back, just waiting for your opinion.

File: torchtitan/distributed/activation_checkpoint.py [hunk 5]
```diff
@@ -297,15 +299,29 @@ def apply_ac(
...
-            transformer_block,
-            ac_config,
-            base_fqn=f"layers.{layer_id}",
-            model_compile_enabled=model_compile_enabled,
-            use_flex_attn=use_flex_attn,
-            op_sac_save_list=op_sac_save_list,
-        )
-        model.layers.register_module(layer_id, transformer_block)
+    if ac_config.mode == "memory_budget":
+        assert (model_compile_enabled is True), "Memory budget mode requires model to be compiled"
```
Comment (tianyu-l) [inline torchtitan/distributed/activation_checkpoint.py]:
```suggestion
        assert model_compile_enabled, "Memory budget mode requires model to be compiled"
```

File: torchtitan/config/job_config.py [hunk 6]
```diff
@@ -566,6 +566,35 @@ class ActivationCheckpoint:
...
+    "flops": Bases it off of the flop count provided by torch.utils.flop_counter
+    "profile": Benchmarks each operator to come up with a runtime
+    "testing": Returns 1 for everything
+    """
+    activation_memory_budget_solver: Literal["dp", "greedy", "ilp"] = "dp"
+    """
+    This controls the solver used for the 0-1 knapsack. By default we use a
+    quantized DP solution ("dp"). The other approaches are a "greedy" and a "ilp"
+    (which has a scipy dependency).
+    """
```
Comment (tianyu-l) [inline torchtitan/config/job_config.py]: It's hard to tell if users should change them. Maybe let's remove them for now and see if people have complaints.

File: torchtitan/config/job_config.py [hunk 7]
```diff
@@ -566,6 +566,35 @@ class ActivationCheckpoint:
...
+    visualize_memory_budget_pareto: bool = False
```
Comment (tianyu-l) [inline torchtitan/config/job_config.py:580]: This one I'm not sure. It could be useful because otherwise people have no idea what value should they set the budge?
Btw, the picture you linked doesn't seem to be in "increments of 0.5". It seems 0.05.

File: torchtitan/config/job_config.py [hunk 8]
```diff
@@ -566,6 +566,35 @@ class ActivationCheckpoint:
     Whether to stop recomputing early when all activations have already been
     rematerialized.
     """
+    activation_memory_budget: float = 1.0
```
Comment (tianyu-l) [inline torchtitan/config/job_config.py]: Maybe set 0.5 by default, o/w if user turns on `memory_budget` without tuning this, nothing would happen.

File: torchtitan/config/job_config.py [hunk 9]
```diff
@@ -566,6 +566,35 @@ class ActivationCheckpoint:
...
+    This controls the solver used for the 0-1 knapsack. By default we use a
+    quantized DP solution ("dp"). The other approaches are a "greedy" and a "ilp"
+    (which has a scipy dependency).
+    """
+    visualize_memory_budget_pareto: bool = False
+    """
+    This dumps out a SVG visualization of the expected runtime vs. activation
+    memory tradeoffs for all memory budget values from 0 to 1 in increments of
+    0.5. See an example here:
+    https://github.com/pytorch/pytorch/pull/126320#discussion_r1625104015
```
Comment (tianyu-l) [inline torchtitan/config/job_config.py:585]: Please describe what folder it'll dump into.

File: torchtitan/config/job_config.py (hunk 8 above)
Comment (tianyu-l) [inline torchtitan/config/job_config.py]: nit: please add space between configs, o/w hard to tell if a message is associated with config above / below.

Comment (tianyu-l) [inline torchtitan/config/job_config.py]: `activation_` is redundant, so we can just call it
```suggestion
    memory_budget: float = 1.0
```

File: torchtitan/config/job_config.py [hunk 10]
```diff
@@ -566,6 +566,35 @@ class ActivationCheckpoint:
...
     rematerialized.
     """
+    activation_memory_budget: float = 1.0
+    """
+    When mode is set to "memory_budget", this value determines how much
+    partitioner in the compiler should trade off compute for memory.
+    0.0 corresponds to the activation memory from applying
+    activation checkpointing to the full compiled region, and 1.0 corresponds to
+    the activation memory from the default runtime-optimized strategy.
+    """
```
Comment (tianyu-l) [inline torchtitan/config/job_config.py]: Please add a link to the post https://pytorch.org/blog/activation-checkpointing-techniques/
